pennylane/devices/qubit_mixed/apply_operation.py

```diff
@@ -0,0 +1,641 @@
+    **_,
+):
+    """Apply an operation to a given state.
+
+    Args:
+        op (Operator): The operation to apply to ``state``
+        state (TensorLike): The starting state.
+        is_state_batched (bool): Boolean representing whether the state is batched or not
+        debugger (_Debugger): The debugger to use
+
+    Keyword Arguments:
+        rng (Optional[numpy.random._generator.Generator]): A NumPy random number generator.
+        prng_key (Optional[jax.random.PRNGKey]): An optional ``jax.random.PRNGKey``. This is
+            the key to the JAX pseudo random number generator. Only for simulation using JAX.
+            If None, a ``numpy.random.default_rng`` will be used for sampling.
+        tape_shots (Shots): the shots object of the tape
+
+    Returns:
+        ndarray: output state
+
+    .. warning::
+
+        ``apply_operation`` is an internal function, and thus subject to change without a deprecation cycle.
+
+    .. warning::
+        ``apply_operation`` applies no validation to its inputs.
+
+        This function assumes that the wires of the operator correspond to indices
+        of the state. See :func:`~.map_wires` to convert operations to integer wire labels.
+
+        The shape of state should be ``[2]*(num_wires * 2)`` (the original tensor form) or
+        ``[2**num_wires, 2**num_wires]`` (the expanded matrix form), where `2`` is
+        the dimension of the system.
+
+    This is a ``functools.singledispatch`` function, so additional specialized kernels
+    for specific operations can be registered like:
+
+    .. code-block:: python
+
+        @apply_operation.register
+        def _(op: type_op, state):
+            # custom op application method here
+
+    **Example:**
+
+    >>> state = np.zeros((2, 2, 2, 2))
+    >>> state[0][0] = 1
+    >>> state
+    array([[[[1., 0.],
+         [0., 0.]],
+
+        [[0., 0.],
+         [0., 0.]]],
+
+
+       [[[0., 0.],
+         [0., 0.]],
+
+        [[0., 0.],
+         [0., 0.]]]])
+    >>> apply_operation(qml.PauliX(0), state)
+    array([[[[0., 0.],
+         [0., 0.]],
+
+        [[0., 0.],
+         [0., 0.]]],
+
+
+       [[[0., 0.],
+         [1., 0.]],
+
+        [[0., 0.],
+         [0., 0.]]]])
+
+    """
+    return _apply_operation_default(op, state, is_state_batched, debugger, **_)
+
+
+def _apply_operation_default(op, state, is_state_batched, debugger, **_):
+    """The default behaviour of apply_operation, accessed through the standard dispatch
+    of apply_operation, as well as conditionally in other dispatches.
+    """
+    if op in diagonal_in_z_basis:
+        return apply_diagonal_unitary(op, state, is_state_batched, debugger, **_)
+    num_op_wires = len(op.wires)
+    interface = math.get_interface(state)
+
+    # Add another layer of condition to rule out batched op (not channel) for tensordot calling
+    mat = op.matrix() + 0j
+    dim = 2**num_op_wires
+    batch_size = math.get_batch_size(mat, (dim, dim), dim**2)
```

Comment:
Could we just use `op.batch_size` here, instead of having to construct the full matrix, only to construct it again later`? Yes, `op.batch_size` isn't entirely reliable when using tensorflow, but I would be ok ignoring that for now.